### PR TITLE
Make storage_client support HTTP

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3783

This is another step towards making HTTP work. The end-to-end integration-y tests in `test_system.py` now run and pass with HTTP storage protocol. At the moment, the HTTP server is disabled by default, see below, so this doesn't affect real-world usage yet. Integration tests still don't pass, [that'll be the next PR hopefully](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3937).

There is a new `tahoe.cfg` option, `[node] force_foolscap = True/False`. At the moment in only affects servers, but it [should be extended to also support clients](https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3936) (**Question for reviewer:** does this mean this should not be under `node`, but some other subsection?). It can be in three states:

* `True`. In this case, only Foolscap is used, HTTP storage protocol is not used.
* `False`. Both Foolscap and HTTP are supported, and for servers NURLs are included in announcements.
* Default, not set. At the moment this translates to same behavior as `True` i.e. just Foolscap is used, but once we're confident HTTP works end-to-end we'll want to change default to be `False`.

Currently on `master` if you start a server it has both HTTP and Foolscap listening. Once this PR lands, it will only listen with HTTP if `force_foolscap` is explicitly set to False in the config, i.e. we're switching back to Foolscap only. Eventually, as mentioned above, we'll want to switch the default to both HTTP and Foolscap listening.

### CODE REVIEW

File: src/allmydata/storage_client.py [hunk 1]
```diff
@@ -684,6 +703,16 @@ def _storage_from_foolscap_plugin(node_config, config, announcement, get_rref):
     raise AnnouncementNotMatched()
 
 
+def _available_space_from_version(version):
```
Comment (itamarst) [inline src/allmydata/storage_client.py:706]: This was moved, unchanged, from lower in the code.

**Comment by itamarst** [comment]
#1226 will fix at least some of the failing tests.

**Comment by itamarst** [comment]
I am investigating test failures.

**Review by exarkun — APPROVED** [review]
Thanks.  This looks pretty good overall.  I left some comments inline.

As far as where the `force_foolscap` config goes ... I'm not sure what the organizing principle behind the config file items is.

Possibly `force_foolscap` should go in `[storage]` since it will control whether a storage server is going to speak Foolscap, GBS, or both.  And it doesn't control the client behavior at all (unless I missed something) so it wouldn't make sense in the `[client]` section.

There are `i2p` and `tor` sections so I suppose there could be an `http` section ... but I don't like that idea very much.

File: src/allmydata/storage/http_client.py [hunk 2]
```diff
@@ -379,21 +401,24 @@ def request(
...
     @inlineCallbacks
     def get_version(self):
         """
         Return the version metadata for the server.
         """
         url = self._client.relative_url("/storage/v1/version")
-        response = yield self._client.request("GET", url)
+        # 1. Getting the version should never take particularly long.
+        # 2. Clients rely on the version command for liveness checks of servers.
+        # Thus, a short timeout.
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_client.py:420]: I feel like there should be a minimal baseline fairly high timeout for everything, which next branch will do. But yeah I can see making `get_version()`'s aggressive timeout happen at higher level, so I'll do that.

File: src/allmydata/storage/http_client.py [hunk 3]
```diff
@@ -276,21 +276,39 @@ def clientConnectionForTLS(self, tlsProtocol):
...
+        test.
+        """
+        cls.TEST_MODE_REGISTER_HTTP_POOL = callback
+
     # The URL is a HTTPS URL ("https://...").  To construct from a NURL, use
     # ``StorageClient.from_nurl()``.
     _base_url: DecodedURL
     _swissnum: bytes
-    _treq: Union[treq, StubTreq, HTTPClient]
+    _treq: Union[treq, StubTreq, HTTPClient] = field(eq=False)
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/storage/http_client.py:304]: I don't agree that it's a misfeature. It's not uncommon to have fields that you don't want impacting equality or hashing, e.g. caches. So mostly I did this out of caution. I will check if removing that breaks anything. And... it's fine.

File: src/allmydata/test/common_system.py [hunk 4]
```diff
@@ -644,18 +641,33 @@ def _render_section_values(values):
...
 class SystemTestMixin(pollmixin.PollMixin, testutil.StallMixin):
 
+    # If set to True, use Foolscap for storage protocol. If set to False, HTTP
+    # will be used when possible. If set to None, this suggests a bug in the
+    # test code.
+    FORCE_FOOLSCAP_FOR_STORAGE : Optional[bool] = None
+
     def setUp(self):
+        self._http_client_pools = []
+        http_client.StorageClient.start_test_mode(self._http_client_pools.append)
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/common_system.py:651]: Uh

**Pull request merged**